File: labs/07-face/python/face-api/face_recognition_service.py
import logging
import os
import time

from azure.cognitiveservices.vision.face import FaceClient
from azure.cognitiveservices.vision.face.models import APIErrorException, DetectionModel, TrainingStatusType
from msrest.authentication import CognitiveServicesCredentials

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class FaceRecognitionService:
    def __init__(self, endpoint, key):
        self.face_client = FaceClient(endpoint, CognitiveServicesCredentials(key))
        self.person_group_id = "myfriends"

    def full_recognition_process(self, image_file_path):
        try:
            self._ensure_person_group_exists()
            self._add_persons_to_group()
            self._train_person_group()
            self._recognize_faces(image_file_path)
        except APIErrorException as e:
            logger.error(f"API error: {e.message} (Code: {e.response.json().get('error', {}).get('code', 'Unknown')})")
            if 'innererror' in e.response.json().get('error', {}):
                inner_error = e.response.json()['error']['innererror']
                logger.error(f"Inner error: {inner_error.get('message')} (Code: {inner_error.get('code')})")
        except Exception as e:
            logger.error("An error occurred during the full recognition process: %s", e)

    def _ensure_person_group_exists(self):
        try:
            self.face_client.person_group.get(person_group_id=self.person_group_id)
            logger.info("Person group already exists.")
        except APIErrorException as e:
            if e.response.status_code == 404:
                self.face_client.person_group.create(person_group_id=self.person_group_id, name="My Friends")
                logger.info("Person group created.")
            else:
                logger.error(f"API error: {e.message}")
                raise e

    def _add_persons_to_group(self):
        try:
            friend1 = self.face_client.person_group_person.create(person_group_id=self.person_group_id, name="Friend1")
            for image_path in os.listdir("images/Aisha"):
                with open(os.path.join("images/Aisha", image_path), 'rb') as image_stream:
                    self.face_client.person_group_person.add_face_from_stream(self.person_group_id, friend1.person_id, image_stream, detection_model=DetectionModel.detection03)
        except APIErrorException as e:
            logger.error(f"API error: {e.message}")
            raise e

    def _train_person_group(self):
        try:
            self.face_client.person_group.train(self.person_group_id)
            logger.info("Training initiated.")
            while True:
                training_status = self.face_client.person_group.get_training_status(self.person_group_id)
                logger.info(f"Training status: {training_status.status}")
                if training_status.status == TrainingStatusType.succeeded:
                    break
                time.sleep(1)
            logger.info("Training completed.")
        except APIErrorException as e:
            logger.error(f"API error: {e.message}")
            raise e

    def _recognize_faces(self, image_file_path):
        try:
            with open(image_file_path, 'rb') as image_stream:
                detected_faces = self.face_client.face.detect_with_stream(image_stream, detection_model=DetectionModel.detection03)
            if not detected_faces:
                logger.info("No faces detected.")
                return

            face_ids = [face.face_id for face in detected_faces]
            identify_results = self.face_client.face.identify(face_ids, self.person_group_id)
            for result in identify_results:
                logger.info(f"Result of face: {result.face_id}")
                if not result.candidates:
                    logger.info("No one identified")
                else:
                    candidate_id = result.candidates[0].person_id
                    person = self.face_client.person_group_person.get(self.person_group_id, candidate_id)
                    logger.info(f"Identified as {person.name}")
        except APIErrorException as e:
            logger.error(f"API error: {e.message}")
            if 'innererror' in e.response.json().get('error', {}):
                inner_error = e.response.json()['error']['innererror']
                logger.error(f"Inner error: {inner_error.get('message')} (Code: {inner_error.get('code')})")
            raise e
        except Exception as e:
            logger.error("An error occurred: %s", e)
            raise e

File: labs/07-face/csharp/face-api/FaceRecognitionService.cs
using Microsoft.Azure.CognitiveServices.Vision.Face;
using Microsoft.Azure.CognitiveServices.Vision.Face.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace Ai102.FaceApi
{
    public class FaceRecognitionService
    {
        private readonly IFaceClient faceClient;
        private const string PersonGroupId = "myfriends";

        public FaceRecognitionService(IFaceClient client)
        {
            faceClient = client;
        }

        public async Task EnsurePersonGroupExists()
        {
            try
            {
                await faceClient.PersonGroup.GetAsync(PersonGroupId);
                Console.WriteLine("Person group already exists.");
            }
            catch (APIErrorException ex) when (ex.Response.StatusCode == System.Net.HttpStatusCode.NotFound)
            {
                await faceClient.PersonGroup.CreateAsync(PersonGroupId, "My Friends");
                Console.WriteLine("Person group created.");
            }
        }

        public async Task AddPersonsToGroup()
        {
            var friend1 = await faceClient.PersonGroupPerson.CreateAsync(PersonGroupId, "Friend1");
            foreach (var imagePath in Directory.GetFiles("images/Aisha"))
            {
                using (var imageStream = File.OpenRead(imagePath))
                {
                    await faceClient.PersonGroupPerson.AddFaceFromStreamAsync(PersonGroupId, friend1.PersonId, imageStream, detectionModel: DetectionModel.Detection03);
                }
            }

            // Repeat for other friends
        }

        public async Task TrainPersonGroup()
        {
            await faceClient.PersonGroup.TrainAsync(PersonGroupId);
            Console.WriteLine("Training initiated.");

            TrainingStatus trainingStatus;
            do
            {
                trainingStatus = await faceClient.PersonGroup.GetTrainingStatusAsync(PersonGroupId);
                Console.WriteLine($"Training status: {trainingStatus.Status}");
                await Task.Delay(1000);
            } while (trainingStatus.Status == TrainingStatusType.Running);

            Console.WriteLine("Training completed.");
        }

        public async Task RecognizeFaces(string imageFilePath)
        {
            IList<DetectedFace> detectedFaces;
            using (var imageStream = File.OpenRead(imageFilePath))
            {
                detectedFaces = await faceClient.Face.DetectWithStreamAsync(imageStream, detectionModel: DetectionModel.Detection03);
            }

            var faceIds = detectedFaces.Select(face => face.FaceId).Where(faceId => faceId.HasValue).Select(faceId => faceId.Value).ToList();
            var identifyResults = await faceClient.Face.IdentifyAsync(faceIds, PersonGroupId);

            foreach (var result in identifyResults)
            {
                Console.WriteLine($"Result of face: {result.FaceId}");
                if (result.Candidates.Count == 0)
                {
                    Console.WriteLine("No one identified");
                }
                else
                {
                    var candidateId = result.Candidates[0].PersonId;
                    var person = await faceClient.PersonGroupPerson.GetAsync(PersonGroupId, candidateId);
                    Console.WriteLine($"Identified as {person.Name}");
                }
            }
        }
    }
}
